internal/migrate/migrator/migrator.go

```diff
@@ -0,0 +1,186 @@
+// everest-operator
+// Copyright (C) 2025 Percona LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package migrator contains binary to migrate resources between versions
+package migrator
+
+import (
+	"context"
+	"fmt"
+	"os"
+
+	"github.com/go-logr/logr"
+	v1 "k8s.io/api/coordination/v1"
+	storagev1 "k8s.io/api/storage/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/config"
+
+	everestv1alpha1 "github.com/percona/everest-operator/api/v1alpha1"
+	"github.com/percona/everest-operator/internal/consts"
+	"github.com/percona/everest-operator/internal/controller/common"
+)
+
+const (
+	systemNamespaceEnvVar              = "SYSTEM_NAMESPACE"
+	versionEnvVar                      = "EVEREST_VERSION"
+	leaseName                          = "everest-migration-lease"
+	lastMigrationVersionAnnotationName = "last-migration-version"
+	defaultEKSLoadBalancerConfigName   = "eks-default"
+)
+
+// Migrator is the struct representing the migrator.
+type Migrator struct {
+	l               logr.Logger
+	currentVersion  string
+	systemNamespace string
+	client          client.Client
+}
+
+// NewMigrator creates a new migrator.
+func NewMigrator(logger logr.Logger) (*Migrator, error) {
+	m := &Migrator{
+		l:               logger,
+		currentVersion:  os.Getenv(versionEnvVar),
+		systemNamespace: os.Getenv(systemNamespaceEnvVar),
+	}
+
+	cl, err := client.New(config.GetConfigOrDie(), client.Options{Scheme: m.BuildScheme()})
+	if err != nil {
+		return nil, fmt.Errorf("unable to create client: %w", err)
+	}
+
+	m.client = cl
+
+	return m, nil
+}
+
+// Migrate function that performs CRs migration.
+func (m *Migrator) Migrate(ctx context.Context) (info string, rerr error) { //nolint:nonamedreturns
+	lease, err := m.getLease(ctx)
+	if err != nil {
+		return "", fmt.Errorf("unable to prepare lease: %w", err)
+	}
+
+	defer func() {
+		if rerr != nil {
+			return
+		}
+
+		lease.Annotations[lastMigrationVersionAnnotationName] = m.currentVersion
+
+		err = m.client.Update(ctx, lease)
+		if err != nil {
+			rerr = fmt.Errorf("unable to update lease: %w", err)
+		}
```

Comment:
I'm hesitant about running this logic inside a defer block. If the lease update fails, the init container will exit successfully as long as the migration itself worked. This means the Everest operator will continue to run, and on the next pod restart, the migration will execute again because the lease wasn't updated.

The core question is whether we want the operator to run if the migration succeeded but the lease update failed, or if we should force it to wait for a successful lease update before proceeding.

In both options, we need to make sure that the migration logic is idempotent because a lease update failure would lead to the migration running again eventually.

Reply:
> If the lease update fails, the init container will exit successfully as long as the migration itself worked. 

I don't think it's true. If the Lease update fails then the whole `Migrate()` function will fail (because in the `defer` we redefine the `rerr` which is being returned from the function). And then since the `Migrate()` function returns an error, `exit(1)` is going to be called and the initContainer will fail. Added a unit-test illustrating the failure

Reply:
You are right, I missed the rerr definition in the function signature.